### PR TITLE
[ios] Support getLeaves (and related) clustering methods

### LINKED ISSUE #12442: Support `getLeaves` and other new clustering methods in the iOS SDK
We need an iOS equivalent of https://github.com/mapbox/mapbox-gl-js/pull/6829 to support features created in the [most recent release of supercluster.hpp](https://github.com/mapbox/supercluster.hpp/releases/tag/v0.3.0) including `getChildren`, `getLeaves`, and `getClusterExpansionZoom`.

### PR DESCRIPTION
**Pull request by julianrex**
Fixes #12442.

~~This WIP PR adds an `MGLPointCluster` protocol, and `- (id<MGLPointCluster>)cluster` to `MGLPointFeature` which returns non-null for clusters. (`MGLPointFeature` conforms to `MGLPointCluster` privately).~~

~~We could also consider these options:~~
- ~~Add the above `cluster` method to `MGLFeature`,  for future compatibility (e.g. if supercluster supported features other than points)~~
- ~~Instead of a protocol, create a subclass `MGLPointFeature` for point clusters, e.g. `MGLPointClusterFeature`. This may be preferable, but again if other features end up being clustered we could suffer an explosion of types.~~

~~Still TBD: documentation, tests, macos project.~~

This PR introduces a new public protocol, `MGLCluster`, that ~~private types (that conform to `MGLFeature`) will also conform to if they have the appropriate cluster attributes.~~ `MGLPointFeatureCluster` (a new subclass of `MGLPointFeature`) conforms to. 

(Currently the only supported type that adopts `MGLCluster` is a subclass of `MGLPointFeature`.)

This also adds:

- `-[MGLShapeSource leavesOfCluster:offset:limit:]`
- `-[MGLShapeSource childrenOfCluster:]`
- `-[MGLShapeSource zoomLevelForExpandingCluster:]`

exposing features [added to supercluster](https://github.com/mapbox/supercluster.hpp/releases/tag/v0.3.0) 
/cc @tobrun

### CODE REVIEW

File: platform/darwin/src/MGLShapeSource.h [hunk 1]
```diff
@@ -305,6 +307,10 @@ MGL_EXPORT
  */
 - (NSArray<id <MGLFeature>> *)featuresMatchingPredicate:(nullable NSPredicate *)predicate;
 
+// TODO: doc
```
Comment (captainbarbosa) [inline platform/darwin/src/MGLShapeSource.h]: When you do write the docs for this part, I think its a good idea to add extra clarification on what are leaves vs. clusters since the metaphor is a bit odd..

File: platform/darwin/src/MGLFeature.mm [hunk 2]
```diff
@@ -51,6 +58,56 @@ - (NSString *)description
...
+}
+
+#pragma mark - MGLPointCluster
+
+- (BOOL)isCluster {
+    NSNumber *isCluster = MGL_OBJC_DYNAMIC_CAST(self.attributes[MGLPointClusterBoolKey], NSNumber);
+    return [isCluster boolValue];
+}
+
+- (uint32_t)clusterId {
```
Comment (friedbunny) [inline platform/darwin/src/MGLFeature.mm]: Can you go into the decision to keep `uint32_t` as the public type for this? I’m concerned about Swift bridging and Obj-C ergonomics.

File: platform/darwin/src/MGLFeature.mm [hunk 3]
```diff
@@ -51,6 +58,56 @@ - (NSString *)description
...
+    
+    NSAssert(pointCount, @"Clusters should have a point_count");
+    if (!pointCount) {
+        return -1;
+    }
+
+    return [pointCount unsignedIntegerValue];
+}
+
+- (NSString*)clusterAbbreviated {
```
Comment (friedbunny) [inline platform/darwin/src/MGLFeature.mm]: “Abbreviated” feels ungrammatical here and I’m unsure what this method is meant to do.

File: platform/darwin/src/MGLShapeSource.h [hunk 4]
```diff
@@ -305,6 +307,10 @@ MGL_EXPORT
...
+- (NSArray<id <MGLFeature>> *)leavesOfCluster:(id<MGLPointCluster>)cluster offset:(uint32_t)offset limit:(uint32_t)limit;
```
Comment (friedbunny) [inline platform/darwin/src/MGLShapeSource.h]: As with above, curious about the necessity of `uint32_t` here.

File: platform/darwin/src/MGLFeature.mm (hunk 2 above)
Comment (1ec5) [reply platform/darwin/src/MGLFeature.mm]: The main deciding factor would be whether a type is capable of representing the maximum cluster ID.

`uint32_t` bridges to Swift as `Int32`, which isn’t the end of the world but also fairly unusual. Normally, the only APIs that use that type are ones that deal with low-level devices such as file handles.

`mbgl::AnnotationID` is similarly `uint32_t`, but `MGLAnnotationTag` is `NSUInteger` instead, for familiarity’s sake. mbgl treats GeoJSON’s number type as `uint64_t` or `int64_t` depending, but the iOS/macOS map SDK uses (unsigned) `long long`, which bridges to Swift as (U)`Int64` regardless.

File: platform/darwin/src/MGLPointCluster.h [hunk 5]
```diff
@@ -0,0 +1,15 @@
+#import <Foundation/Foundation.h>
+#import "MGLFoundation.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+MGL_EXPORT
+// TODO: Should this be a MGLPointClusterFeature? Subclass of MGLPointFeature?
+@protocol MGLPointCluster <NSObject>
+@property (nonatomic, readonly, getter=isCluster) BOOL cluster;     // Keep so we match cluster def? Always YES.
+@property (nonatomic, readonly) uint32_t clusterId;
```
Comment (1ec5) [inline platform/darwin/src/MGLPointCluster.h]: Spell out “identifier” in symbol names, rather than abbreviating as `ID` or `Id`.

File: platform/darwin/src/MGLFeature.mm (hunk 2 above)
Comment (julianrex) [reply platform/darwin/src/MGLFeature.mm]: How about we have a typedef? Something like `MGLClusterIdentifier`?

File: platform/darwin/src/MGLFeature.mm (hunk 3 above)
Comment (julianrex) [reply platform/darwin/src/MGLFeature.mm]: It's really the *point count* represented in a text friendly form. I agree that the name is unpleasant, but it currently matches the underlying data & JS ("cluster_abbreviated"). How do you feel about deviating from that? If so, we could come up with something a lot clearer 🙂 

For that matter, we could address the mixed-metaphor of `leaves` vs `children` - though if we do that we'd probably want to do the same on Android. What do you think @tobrun ?

File: platform/darwin/src/MGLShapeSource.h (hunk 4 above)
Comment (julianrex) [reply platform/darwin/src/MGLShapeSource.h]: These I think we could change without too much concern.

**Comment by julianrex** [comment]
@1ec5 I think that's a great goal - I'm hoping that a simpler interface along the lines of MapKit's can be something that can follow after and inform this PR. /cc @chloekraw

File: platform/darwin/src/MGLFeature.mm (hunk 2 above)
Comment (1ec5) [reply platform/darwin/src/MGLFeature.mm]: Note that a typedef has implications for Swift code: you have to cast even between MGLClusterIdentifier and the raw type. Perhaps that extra type safety is warranted; perhaps not. But the more important question @friedbunny raises is about the raw data type: what are the minimum and maximum values that this identifier can conceivably be set to without underflowing or overflowing?

Comment (julianrex) [reply platform/darwin/src/MGLFeature.mm]: > what are the minimum and maximum values that this identifier can conceivably be set to without underflowing or overflowing, and what types are most suitable for those limits in Objective-C and Swift?

32 bits. `UInt32` would be the more Obj-C choice here.

Comment (1ec5) [reply platform/darwin/src/MGLFeature.mm]: `NSUInteger` and `UInt` are more ergonomic than `uint32_t` and `UInt32` in Objective-C and Swift, respectively (and the signed types are more ergonomic than the unsigned types). So as long as those more ergonomic types can fit all the possible values, then it’s OK even if the type isn’t exactly the same width.

Hypothetically, if a public read-write property uses a wider type than the underlying internal type, then the getter and application have nothing to worry about, but the setter needs some bounds checking. On the other hand, if the property uses the same type as the underlying code but it’s narrower than the more common type used in application code (such as `NSUInteger`), then the application ends up needing to bridge the impedance mismatch with explicit casts – that can be pretty error prone.

> Update: Interestingly it looks like `MGLAnnotationTag` is a `uint32_t` after all.

Ah, right. Perhaps of interest, `MGLAnnotationTagNotFound` is `UINT32_MAX`, under the assumption that `UINT32_MAX` − 1 is enough annotations for an uncontrived application.

https://github.com/mapbox/mapbox-gl-native/blob/44aa4f4be2bd504da982ffdcc0ffa5e84b2c6f31/platform/macos/src/MGLMapView.mm#L107

**Comment by julianrex** [comment]
@tobrun I've updated this PR with latest changes - would you mind checking to see if this matches what you expect?

File: platform/darwin/src/MGLShapeSource.h [hunk 6]
```diff
@@ -321,6 +323,49 @@ MGL_EXPORT
...
+ the `MGLCluster` protocol.
+ 
+ @param cluster An object that conforms to the `MGLCluster` protocol. Currently
+    the only types that can conform are private subclasses of `MGLPointFeature`.
+ 
+ @return An array of objects that conform to the `MGLFeature` protocol.
+ 
+ @note The returned array may contain the `cluster` that was passed in, if the next
+    zoom level doesn't the zoom level for expanding that cluster. See
+    `-[MGLShapeSource zoomLevelForExpandingCluster:`.
```
Comment (friedbunny) [inline platform/darwin/src/MGLShapeSource.h]: This method name is missing a closing `]`.

File: platform/darwin/test/MGLDocumentationExampleTests.swift [hunk 7]
```diff
@@ -374,7 +374,7 @@ class MGLDocumentationExampleTests: XCTestCase, MGLMapViewDelegate {
                 return MGLDocumentationExampleTests.styleURL
             }
         }
-        
+
```
Comment (friedbunny) [inline platform/darwin/test/MGLDocumentationExampleTests.swift]: Uber-nit: unnecessary whitespace change.

File: platform/darwin/src/MGLShapeSource.h (hunk 6 above)
Comment (julianrex) [reply platform/darwin/src/MGLShapeSource.h]: Fixed.

File: platform/darwin/test/MGLDocumentationExampleTests.swift (hunk 7 above)
Comment (julianrex) [reply platform/darwin/test/MGLDocumentationExampleTests.swift]: Fixed?

File: platform/darwin/src/MGLShapeSource.h [hunk 8]
```diff
@@ -321,6 +323,49 @@ MGL_EXPORT
...
+ and a maximum number of features to return.
+ 
+ @param cluster An object that conforms to the `MGLCluster` protocol. Currently
+    the only types that can conform are private subclasses of `MGLPointFeature`.
+ @param offset Number of features to skip.
+ @param limit Maximum number of features to return
+ 
+ @return An array of objects that conform to the `MGLFeature` protocol.
+ */
+- (NSArray<id <MGLFeature>> *)leavesOfCluster:(id<MGLCluster>)cluster offset:(NSUInteger)offset limit:(NSUInteger)limit;
```
Comment (1ec5) [inline platform/darwin/src/MGLShapeSource.h]: MapKit’s API refers to a cluster’s constitutent features as “members”. I think “member” is more self-explanatory than “leaf”; the analogy to trees requires some explanation.

File: platform/darwin/src/MGLShapeSource.h [hunk 9]
```diff
@@ -321,6 +323,49 @@ MGL_EXPORT
...
+ Returns an array of map features that are the leaves of the specified cluster.
+ ("Leaves" are the original points that belong to the cluster.)
+ 
+ This method supports pagination; you supply an offset (number of features to skip)
+ and a maximum number of features to return.
+ 
+ @param cluster An object that conforms to the `MGLCluster` protocol. Currently
+    the only types that can conform are private subclasses of `MGLPointFeature`.
+ @param offset Number of features to skip.
+ @param limit Maximum number of features to return
```
Comment (1ec5) [inline platform/darwin/src/MGLShapeSource.h]:
```suggestion
 @param limit The maximum number of features to return.
```

File: platform/darwin/src/MGLShapeSource.h [hunk 10]
```diff
@@ -321,6 +323,49 @@ MGL_EXPORT
...
+ */
+- (NSArray<id <MGLFeature>> *)leavesOfCluster:(id<MGLCluster>)cluster offset:(NSUInteger)offset limit:(NSUInteger)limit;
+
+/**
+ Returns an array of map features that are the immediate children of the specified
+ cluster *on the next zoom level*. The may include features that also conform to
+ the `MGLCluster` protocol.
+ 
+ @param cluster An object that conforms to the `MGLCluster` protocol. Currently
+    the only types that can conform are private subclasses of `MGLPointFeature`.
```
Comment (1ec5) [inline platform/darwin/src/MGLShapeSource.h]: I’m not sure developers intuitively understand the implications of a private subclass. How about:

> This object should be an object returned from a feature querying method, such as `-[MGLMapView visibleFeaturesInRect:]` or `-[MGLShapeSource featuresMatchingPredicate:]`. Do not pass in a custom class that conforms to this protocol.

File: platform/darwin/src/MGLCluster.h [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+ An `NSUInteger` constant used to indicate an invalid cluster identifier.
+ This indicates a missing cluster feature.
+ */
+FOUNDATION_EXTERN MGL_EXPORT const NSUInteger MGLClusterIdentifierInvalid;
+
+/**
+ A protocol that (private) feature subclasses (i.e. those already conforming to
+ the `MGLFeature` protocol) conform to if they are represent clusters.
+ 
+ Currently only subclasses of `MGLPointFeature` support `MGLCluster`.
```
Comment (1ec5) [inline platform/darwin/src/MGLCluster.h]: This only says that we aren’t defining additional classes that conform to the MGLCluster protocol. However, it doesn’t dissuade developers from defining their own classes that conform to the protocol and creating instances of those classes, which is technically possible and might be tempting under some circumstances.

I’m curious why we wound up making this a protocol instead of a subclass of MGLPointFeature. Is it expected that clusters of non-point features could be inspected in the future?

File: platform/darwin/src/MGLCluster.h [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+ // Check for cluster conformance
+ guard let cluster = pointFeature as? MGLCluster else {
+     throw ExampleError.featureIsNotACluster
+ }
+ 
+ print("Approximate number of points in cluster: \(cluster.clusterPointCountAbbreviation)")
+ ```
+ */
+MGL_EXPORT
+@protocol MGLCluster <NSObject>
```
Comment (1ec5) [inline platform/darwin/src/MGLCluster.h]: Unless we expect non-features to be clusterable in the future, let’s make this protocol conform to MGLFeature.

```suggestion
@protocol MGLCluster <MGLFeature>
```

File: platform/darwin/src/MGLFeature_Private.h [hunk 13]
```diff
@@ -46,6 +47,10 @@ NS_ASSUME_NONNULL_END
             NSSet<Class> *identifierClasses = [NSSet setWithArray:@[[NSString class], [NSNumber class]]]; \
             identifier = [decoder decodeObjectOfClasses:identifierClasses forKey:@"identifier"]; \
             attributes = [decoder decodeObjectOfClass:[NSDictionary class] forKey:@"attributes"]; \
+            /* If this feature is a cluster, then change the class */ \
+            if (MGLFeatureHasClusterAttribute(self)) { \
+                MGLConvertFeatureToClusterSubclass(self); \
```
Comment (1ec5) [inline platform/darwin/src/MGLFeature_Private.h]: It feels pretty hacky to change the class within an initializer after calling the superclass’s initializer. In principle, this should only be necessary if we expect that a non-MGLCluster could be archived then unarchived as an MGLCluster, which seems unlikely.

File: platform/darwin/src/MGLFeature.mm [hunk 14]
```diff
@@ -442,6 +443,10 @@ static CLLocationCoordinate2D toLocationCoordinate2D(const mbgl::Point<T> &point
     }
     shape.attributes = attributes;
 
+    if (MGLFeatureHasClusterAttribute(shape)) {
+        MGLConvertFeatureToClusterSubclass(shape);
```
Comment (1ec5) [inline platform/darwin/src/MGLFeature.mm]: Consider modifying `GeometryEvaluator` to create an MGLCluster subclass instance directly:

https://github.com/mapbox/mapbox-gl-native/blob/0addd511d5fa503dc67d0773893c66c000e3ec98/platform/darwin/src/MGLFeature.mm#L320

File: platform/darwin/src/MGLFeature.mm (hunk 3 above)
Comment (julianrex) [reply platform/darwin/src/MGLFeature.mm]: Have removed the abbreviation property from latest rebase/push.

**Comment by julianrex** [comment]
This PR is now the equivalent of the Android PR (https://github.com/mapbox/mapbox-gl-native/pull/13631) due to the introduction of the `queryFeatureExtension` API introduced in https://github.com/mapbox/mapbox-gl-native/pull/13382.

Because of the introduction of this API, this PR has been refactored so that an `MGLShape` that conforms to `MGLCluster` is used: specifically this is `MGLPointFeatureCluster` (a concrete subclass of `MGLPointFeature`). `MGLCluster has not changed` other than the static free functions that were used to create cluster subclasses at runtime being turned in to methods of `MGLPointFeatureCluster`.

If it turns out that types other than MGLPointFeature can be clustered, we can revisit. Though I suspect that we'll want to address #7454 before then.

File: platform/darwin/src/MGLFeature.h [hunk 15]
```diff
@@ -186,12 +187,27 @@ MGL_EXPORT
...
+ An `MGLPointFeatureCluster` object associates a point shape (with an optional
+ identifier and attributes) and represents a point cluster.
+ 
+ @see `MGLCluster`
+ 
+ #### Related examples
+ See the <a href="https://www.mapbox.com/ios-sdk/maps/examples/clustering/">
+ Clustering point data</a> example to learn how to initialize
+ clusters and add them to your map.
+ */
```
Comment (friedbunny) [inline platform/darwin/src/MGLFeature.h]: screen shot 2019-01-14 at 12 41 27 pm

Jazzy currently puts this class (and `MGLEmptyFeature`) into the “Other Classes” category — to customize this, edit [`jazzy.yml`](https://github.com/mapbox/mapbox-gl-native/blob/master/platform/ios/jazzy.yml).

File: platform/darwin/src/MGLFeature.mm [hunk 16]
```diff
@@ -374,7 +430,7 @@ - (NSDictionary *)geoJSONDictionary {
         return [MGLShapeCollectionFeature shapeCollectionWithShapes:shapes];
     }
 
-private:
+private:
```
Comment (friedbunny) [inline platform/darwin/src/MGLFeature.mm]: Nit: new whitespace.

File: platform/darwin/src/MGLShapeSource.h (hunk 9 above)
Comment (friedbunny) [reply platform/darwin/src/MGLShapeSource.h]: ☝️

File: platform/darwin/test/MGLCodingTests.mm [hunk 17]
```diff
@@ -41,6 +44,36 @@ - (void)testPointFeature {
...
+- (void)testPointFeatureCluster {
+    MGLPointFeature *pointFeature = [[MGLPointFeatureCluster alloc] init];
+    pointFeature.title = @"title";
+    pointFeature.subtitle = @"subtitle";
+    pointFeature.identifier = @(123);
+    pointFeature.attributes = @{
+        @"cluster" : @(YES),
+        @"cluster_id" : @(456),
+        @"point_count" : @(2),
+        };
```
Comment (friedbunny) [inline platform/darwin/test/MGLCodingTests.mm]: Super-nit: the indentation here is SLIGHTLY OFF. 🙅‍♂️

File: platform/darwin/src/MGLFeature.h (hunk 15 above)
Comment (julianrex) [reply platform/darwin/src/MGLFeature.h]: Perfect - thanks for this.

File: platform/darwin/src/MGLFeature.mm (hunk 16 above)
Comment (julianrex) [reply platform/darwin/src/MGLFeature.mm]: Done

File: platform/darwin/test/MGLCodingTests.mm (hunk 17 above)
Comment (julianrex) [reply platform/darwin/test/MGLCodingTests.mm]: Done

File: platform/darwin/src/MGLCluster.h [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+
+/** The identifier for the cluster. */
+@property (nonatomic, readonly) NSUInteger clusterIdentifier;
+
+/** The number of points within this cluster */
+@property (nonatomic, readonly) NSUInteger clusterPointCount;
+
+/**
+ An `NSString` abbreviation for the number of points within this cluster. For
+ example "1.2k".
```
[the remark this answers is not in the document]
Comment (samrayner) [reply platform/darwin/src/MGLCluster.h]: Sorry to randomly comment on this 2 years later but in case anyone stumbles across this while researching how to abbreviate cluster counts on iOS, 1ec5's code got me 90% of the way there:

```
clusterNumbersLayer.text = NSExpression(
            format: """
                    TERNARY(point_count >= 1000,
                        mgl_join({
                            CAST(
                                TERNARY(point_count >= 10000,
                                    mgl_round(point_count / 1000),
                                    mgl_round(point_count / 100) / 10
                                ),
                                'NSString'
                            ),
                            'k'
                        }),
                        point_count
                    )
                    """
        )
```
